sdk/cosmosdb/cosmos/src/request/CosmosDiagnostics.ts

```diff
@@ -0,0 +1,212 @@
+  id: string;
+  startTimeUTC: number;
+  endTimeUTC: number;
+  metaDataType: MetadataType;
+  activityId: string;
+}
+
+/**
+ * This type captures diagnostic information regarding a failed request to server api.
+ */
+export interface FailedRequestAttempt {
+  id: string;
+  attemptNumber: number;
+  startTimeUTC: number;
+  endTimeUTC: number;
+  statusCode: string;
+}
+
+export enum MetadataType {
+  CONTAINER_LOOK_UP,
+  PARTITION_KEY_RANGE_LOOK_UP,
+  SERVER_ADDRESS_LOOKUP,
+  DATABASE_ACCOUNT_LOOKUP,
+}
+
+export type ClientSideRequestStatistics = {
+  requestStartTimeUTC: number;
+  requestEndTimeUTC: number;
+  activityId: string;
+  locationEndpointsContacted: string[];
+  retryDiagnostics: RetryDiagnostics;
+  metadataDiagnostics: MetadataDiagnostics;
+  requestPayloadLength: number;
+  responsePayloadLength: number;
+};
+
+export function getCurrentTimestamp(): number {
+  return Date.now();
+}
+
+/**
+ * Utility function to create an Empty CosmosDiagnostic object.
+ * @returns
+ */
+export function getEmptyCosmosDiagnostics(): CosmosDiagnostics {
+  return {
+    id: v4(),
+    clientSideRequestStatistics: {
+      activityId: "",
+      requestEndTimeUTC: 0,
+      requestStartTimeUTC: 0,
+      locationEndpointsContacted: [],
+      retryDiagnostics: {
+        failedAttempts: [],
+      },
+      metadataDiagnostics: {
+        metadataLookups: [],
+      },
+      requestPayloadLength: 0,
+      responsePayloadLength: 0,
+    },
+  };
+}
+/**
+ * @hidden
+ * Internal class to hold CosmosDiagnostic information all through the lifecycle of a request.
+ * This object gathers diagnostic information throughout Client operation which may span across multiple
+ * Server call, retries etc.
+ * Functions - recordFailedAttempt, recordMetaDataQuery, recordEndpointContactEvent are used to ingest
+ * data into the context. At the end of operation, getDiagnostics() is used to
+ * get final CosmosDiagnostic object.
+ */
+export class CosmosDiagnosticContext {
+  private requestStartTimeUTC: number;
+  private requestEndTimeUTC: number;
+  private retryStartTimeUTC: number;
+  private headers: CosmosHeaders = {};
+  private retryAttempNumber: number;
+  private failedAttempts: FailedRequestAttempt[] = [];
+  private locationEndpointsContacted: Set<string> = new Set();
+  private metadataLookups: MetadataLookup[] = [];
+  private requestPayloadLength: number;
+  private responsePayloadLength: number;
+  public constructor() {
+    this.requestStartTimeUTC = getCurrentTimestamp();
+    this.retryStartTimeUTC = this.requestStartTimeUTC;
+  }
+
+  public recordRequestPayload(payload: string): void {
+    this.requestPayloadLength = payload.length;
+  }
+
+  public recordResponseStats(payload: string, headers: CosmosHeaders): void {
+    this.responsePayloadLength = typeof payload === "string" ? payload.length : 0;
+    this.headers = { ...this.headers, ...headers };
+  }
+
+  public recordFailedAttempt(statusCode: string): void {
```

Comment:
Can you pls elaborate on this? This is an internal function. (hence we have strict control of which functions call it) + meta operation (Lower in priority then the actual resource calls)

Reply:
It's alright, @v1k1 , but currently, we're storing the status code directly in FailedRequestAttemptDiagnostic. Would it be possible to add some validation for the status code so that only a specific set of values can be allowed?

Reply:
Hey @topshot99, we are ourselves constructing the status code, so checking it again inside diagnostics seems redundant.